lib/driver_sht41/sht41.c

```diff
@@ -2,118 +2,166 @@
 #define SHT41_POLL_TIME_MS 500
 #define SHT41_INVALID_TIMEOUT_MS 600
 
-static void on_write(const struct async_i2c_request *req);
+// Global varibles
+volatile bool temp_rh_set_on_read = false;
+
+// fnc prototypes
+static int64_t sht41_start_write_req(__unused alarm_id_t id, void *user_data);
+
+static void sht41_on_write(const struct async_i2c_request *req);
+
+static int64_t sht41_start_read_req(__unused alarm_id_t id, void *user_data);
+
+static void sht41_on_read(const struct async_i2c_request *req);
 
-static void on_read(const struct async_i2c_request *req);
+// static void i2c_error_complaint(__unused const struct async_i2c_request *req, uint32_t error_code);
 
-static void on_error();
+static void sht41_on_error(const sht41_error_code error_type, uint32_t i2c_error_code);
 
+// SHT41 data fields
 static const uint8_t write_cmd[] = { SHT41_WRITE_CMD };
 static uint8_t rx_buf[SHT41_READ_LENGTH];
-// static struct async_i2c_request read_req = { .i2c_num = 0,  // FIXME put in i2c_num
-//                                              .address = SHT41_I2C_ADDR,
-//                                              .nostop = false,
-//                                              .tx_buffer = NULL,
-//                                              .rx_buffer = rx_buf,
-//                                              .bytes_to_send = 0,
-//                                              .bytes_to_receive = SHT41_READ_LENGTH,
-//                                              .completed_callback = on_read,
-//                                              .failed_callback = on_error,
-//                                              .next_req_on_success = NULL };
+
+// requests
+static struct async_i2c_request wr_req = {
+    .i2c_num = BOARD_I2C,
+    .address = SHT41_I2C_ADDR,
+    .nostop = false,
+    .tx_buffer = write_cmd,
+    .rx_buffer = NULL,
+    .bytes_to_send = 1,
+    .bytes_to_receive = 0,
+    .completed_callback = &sht41_on_write,
+    .failed_callback = &on_async_i2c_error,
+    .next_req_on_success = NULL,
+    .user_data = NULL  // TODO double check one this one
+};
+
+static struct async_i2c_request rd_req = {
+    .i2c_num = BOARD_I2C,
+    .address = SHT41_I2C_ADDR,
+    .nostop = false,
+    .tx_buffer = NULL,
+    .rx_buffer = rx_buf,
+    .bytes_to_send = 0,
+    .bytes_to_receive = 6,
+    .completed_callback = &sht41_on_read,
+    .failed_callback = &on_async_i2c_error,
+    .next_req_on_success = NULL,
+    .user_data = NULL  // FIXME double check on this one
+};
 
 static struct sht41_data {
     int16_t temp;
     int16_t rh;
-    absolute_time_t reading_expiration;
-    struct async_i2c_request write_read_req;  // TODO nned to init in the init()
-} last_reading;
+    absolute_time_t reading_expiration;  // TODO nned to init in the init()
+} last_reading = { 0 };
 
 static bool msg_in_progress = false;
 
 static sht41_error_cb error_cb;
 
-// TODO: Add error string
+static uint8_t err_count = 2;
```

Comment:
Make this depend on a constant define, since if you change the # of acceptable errors, this will be invalid